apps/aecore/src/aec_conductor.erl

```diff
@@ -979,26 +980,46 @@ handle_post_block(Block, State) ->
     end.
 
 handle_mined_block(Block, State) ->
-    epoch_mining:info("Block mined: Height = ~p; Hash = ~s",
+    epoch_mining:info("Block mined: Height = ~p; Hash = ~p",
                       [aec_blocks:height(Block),
-                       as_hex(aec_blocks:root_hash(Block))]),
+                       ok(aec_blocks:hash_internal_representation(Block))]),
     handle_add_block(Block, State, block_created).
 
 handle_signed_block(Block, State) ->
-    epoch_mining:info("Block signed: Height = ~p; Hash = ~s",
+    epoch_mining:info("Block signed: Height = ~p; Hash = ~p",
                       [aec_blocks:height(Block),
-                       as_hex(aec_blocks:root_hash(Block))]),
+                       ok(aec_blocks:hash_internal_representation(Block))]),
     handle_add_block(Block, State, micro_block_created).
 
 as_hex(S) ->
     [io_lib:format("~2.16.0b", [X]) || <<X:8>> <= S].
 
+ok({ok, Value}) ->
+    Value;
+ok(Other) ->
+    Other.
+
+
 handle_add_block(Block, #state{} = State, Origin) ->
     Header = aec_blocks:to_header(Block),
     handle_add_block(Header, Block, State, Origin).
 
 handle_add_block(Header, Block, State, Origin) ->
     {ok, Hash} = aec_headers:hash_header(Header),
+    Prev = aec_headers:prev_hash(Header),
+    case Prev =:= State#state.top_block_hash of
+        false when Origin =:= block_created;
+                   Origin =:= micro_block_created;
+                   Origin =:= block_received;
+                   Origin =:= micro_block_received ->
+            epoch_mining:info("Mined block has old top - discarding (~p)", [Prev]),
```

Comment:
We _only_ want to discard work if it is old because of something _we_ did! Not in other cases. This looks wrong.

If we have a problem in testing this isn't the way to solve it...

Reply:
@hanssv was that objection in reference to including `block_received` and `micro_block_received`?

I agree that those checks have no bearing on the problems observed during testing.

Anyway, I will continue to study the issue and see if I can come up with something better.